src/test/java/bio/overture/ego/controller/AppJWTTest.java

```diff
@@ -0,0 +1,231 @@
+    tokenService.isValidToken(accessToken);
+    val accessTokenScope = tokenService.getAppAccessToken(accessToken).getScope();
+
+    // assert jwt scope matches scopes from app permissions
+    assertEquals(scopes, accessTokenScope);
+  }
+
+  @Test
+  @SneakyThrows
+  public void applicationAndGroupPerms_appJwtHasAllResolvedScopes_Success() {
+    val app =
+        entityGenerator.setupApplication("TestCombinedApp", "testsecret", ApplicationType.CLIENT);
+    val group = entityGenerator.setupGroup("Test Group");
+    val policies = entityGenerator.setupPolicies("SONG", "SCORE");
+
+    val appId = app.getId();
+    val groupId = group.getId();
+    val policyId1 = policies.get(0).getId();
+    val policyId2 = policies.get(1).getId();
+
+    // associate app with group
+    val assocGroupReq =
+        initStringRequest().endpoint("/groups/%s/applications", groupId).body(asList(appId)).post();
+    assertEquals(assocGroupReq.getStatusCode(), OK);
+
+    val appPerm1 =
+        initStringRequest()
+            .endpoint("/policies/%s/permission/application/%s", policyId1, appId)
+            .body(createMaskJson(READ.toString()))
+            .post();
+
+    val appPerm2 =
+        initStringRequest()
+            .endpoint("/policies/%s/permission/application/%s", policyId2, appId)
+            .body(createMaskJson(READ.toString()))
+            .post();
+
+    val groupPerm1 =
+        initStringRequest()
+            .endpoint("/policies/%s/permission/group/%s", policyId1, groupId)
+            .body(createMaskJson(WRITE.toString()))
+            .post();
+
+    // get app permissions
+    val appPermsReq = initStringRequest().endpoint("/applications/%s/permissions", appId).getAnd();
+    appPermsReq.assertOk();
+    appPermsReq.assertPageResultHasSize(ApplicationPermission.class, 2);
+
+    // get group permissions
+    val groupPermsReq = initStringRequest().endpoint("/groups/%s/permissions", groupId).getAnd();
+    groupPermsReq.assertOk();
+    groupPermsReq.assertPageResultHasSize(GroupPermission.class, 1);
+
+    // get resolved permissions for app
+    val resolvedAppPermsReq =
+        initStringRequest().endpoint("/applications/%s/groups/permissions", appId).getAnd();
+    resolvedAppPermsReq.assertOk();
+
+    val responseBody = resolvedAppPermsReq.getResponse().getBody();
+    assertNotNull(responseBody);
+    val responseJson = MAPPER.readTree(responseBody);
+
+    // assert there are only 2 resolved permissions because of overlap
+    assertTrue(responseJson.isArray());
+    assertEquals(responseJson.size(), 2);
+
+    val resolvedScopes =
+        Streams.stream(responseJson)
+            .map(
+                x -> {
+                  // this logic assumes result can be only App or Group permission type
+                  val owner = x.get("ownerType").asText();
+                  if (owner.equalsIgnoreCase("group")) {
+                    return gson.fromJson(x.toString(), GroupPermission.class);
```

Comment:
1. rewrite this in light of the previous comment
2. avoid the use of gson json serialization library. We already depend heavily on the Jackson library, and dont need a dependency on another one.